### PR TITLE
Add blockTracker struct

### PR DESCRIPTION
**Pull request by neekolas**
## TL;DR
Added a BlockTracker component to persistently track the latest processed block for each contract address.

### What changed?
- Created a new `BlockTracker` type that maintains the latest processed block number for contract addresses
- Added SQL queries and database table to persist block numbers
- Implemented thread-safe block number updates with atomic operations and mutex locks
- Added comprehensive test coverage for single and concurrent operations
- Created database migration scripts for the new `latest_block` table

### How to test?
- Run the unit tests in `blockTracker_test.go`
- Tests cover:
  - Basic initialization and updates
  - Concurrent updates from multiple goroutines
  - Multiple contract address tracking
  - Persistence across tracker instances
  - Edge cases like updating to lower block numbers

### Why make this change?
This change enables reliable tracking of processed blockchain blocks per contract, preventing duplicate processing and ensuring no blocks are missed during indexing operations. The persistent storage allows the system to resume from the correct block after restarts.

## Summary by CodeRabbit

- **New Features**
  - Introduced new SQL commands to manage the latest block for contract addresses, allowing for efficient retrieval and updating of block numbers.
  - Added a `BlockTracker` functionality to monitor and update the latest indexed block for specific contract addresses.

- **Bug Fixes**
  - Ensured thread safety during concurrent updates to the latest block number.

- **Tests**
  - Implemented comprehensive unit tests for the `BlockTracker` to validate its functionality and performance under various conditions.

- **Migrations**
  - Created a new database table `latest_block` to store contract addresses and their corresponding block numbers, along with a command to drop the table if it exists.

### CODE REVIEW

File: pkg/indexer/blockTracker.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+	bt.mu.Lock()
+	defer bt.mu.Unlock()
+
+	// Re-check after acquiring lock
+	if block <= bt.latestBlock.Load() {
+		return nil
+	}
+
+	if err := bt.updateDB(ctx, block); err != nil {
```
Comment (neekolas) [inline pkg/indexer/blockTracker.go:64]: I didn't want to over-think it for now, but I may end up debouncing updates to the DB so that it doesn't update on every block when you are doing a historical sync

**Review by richardhuaaa — APPROVED** [review]

**Comment by neekolas** [comment]
### Merge activity

* **Nov 25, 7:46 AM PST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/xmtp/xmtpd/283).
